Add Users API endpoint

Khristophor opened the pull request:
TODO:
- Authentication; currently anyone can hit `/api/users/` or `/api/users/{id}` to grab user info
- Integration into the React portion for areas where we need User info
- Refactoring of the Manager area to defer to the User model for things like first/last name

Khristophor commented on resources/assets/js/models/types.ts:
```diff
@@ -47,6 +160,25 @@ export interface Job {
+  work_env_description: string | null;
+  culture_summary: string | null;
+  culture_special: string | null;
+}
```
Ah my bad, I was focused on not losing things during the merge

Khristophor commented on app/Http/Controllers/Api/UserController.php:
```diff
@@ -0,0 +1,63 @@
+                return $this->unauthorized;
+            }
+            return $next($request);
+        });
```
@tristan-orourke So I think I skipped that one because we're using a custom `redirectTo()` in our auth middleware that just redirects the user to a login page if they're not authenticated. Unless it doesn't fire for API requests?

tristan-orourke replied:
That's a good point. I'm not sure how it works for ajax requests.

Khristophor replied:
This is the kind of thing I was concerned about having API routes/controllers mixed in with the rest of the web routes... It's going to get more complicated and convoluted mixing up request and response types and attempting to use the same middleware everywhere.

Laravel has conventions in place for handling web and api requests separately, and I think we should be refactoring towards that in the future.

tristan-orourke replied:
Good point. If we were to set up a different auth guard for the api routes, the 'auth' middleware might work more appropriately for them.

Khristophor commented on routes/web.php:
```diff
@@ -643,6 +643,9 @@ function (): void {
     // Protected by a gate in the controller, instead of policy middleware.
     Route::get('jobs/{jobPoster}/assessment-plan', 'AssessmentPlanController@getForJob');
+    Route::get('users', 'Api\UserController@index');
+    Route::get('users/{user}', 'Api\UserController@show')
```
Wouldn't I have to register it in `Kernel.php` with an associated Middleware file though? Right now it's a gate.

tristan-orourke replied:
I got the syntax wrong in my suggestion, it would be `->middleware('can:view-user', 'user')`. So you wouldn't have to register a new middleware, the 'can' middleware should work with any Gate or Policy.

Khristophor commented:
@tristan-orourke made a handful of changes:

- Updated `Handler.php`... a lot of the classes in `internalDontReport` weren't namespaced properly or even seemed to exist anymore. I also added a `wantsJson`/`AuthorizationException` check to return a json response as it was always returning a 403 HTML error page if `$this->authorize()` failed in the controller.
- Changed out my json unauthorized response for throwing an AuthorizationException to keep things consistent with the rest of the app
- I elected not to call the gate from the routes file since it's using an `authorize()` check in the controller, having both is redundant

tristan-orourke commented on app/Http/Controllers/Api/UserController.php:
```diff
@@ -0,0 +1,59 @@
+     */
+    public function index()
+    {
+        $users = User::with(['applicant', 'manager', 'hr_advisor'])->get();
```
Do we need the profile data to come back with the user? If we're not using it, I recommend not loading it. Keeping the profiles separate keeps the permissions simpler, conceptually.

Khristophor replied:
I am going to have to include them in the query here for the sake of checking Department permissions in the Gate, but this can be cleaned up once #2846 is finished. I can exclude it from the response thanks to the Laravel API Resource classes.
